### PR TITLE
Insertion of EntryPoints into modeling.fitters.

### PR DESCRIPTION
**Pull request by nocturnalastro**
This allows insertion of external references into the astropy.modeling.fitting namespace. This is required by the google summer of code [Saba project](https://github.com/nocturnalastro/saba) which allows the astropy users access to sherpa's fitting routines.  It would also be the mechanism for supporting other external back-end fitters such as emcee.

### CODE REVIEW

File: CHANGES.rst [hunk 1]
```diff
@@ -59,6 +59,11 @@ New Features
 
 - ``astropy.wcs``
 
+  - wcslib was updated to v5.14 [#4579]
+
+- Added the injection of EntryPoints into astropy.modeling.fitting if they inherit from Fitters class.
+
+API changes
```
Comment (pllim) [inline CHANGES.rst]: Remove duplicate line here.

File: CHANGES.rst [hunk 2]
```diff
@@ -59,6 +59,11 @@ New Features
 
 - ``astropy.wcs``
 
+  - wcslib was updated to v5.14 [#4579]
+
+- Added the injection of EntryPoints into astropy.modeling.fitting if they inherit from Fitters class.
```
Comment (pllim) [inline CHANGES.rst]: Add PR number. See other entries for example.

File: CHANGES.rst [hunk 3]
```diff
@@ -59,6 +59,11 @@ New Features
 
 - ``astropy.wcs``
 
+  - wcslib was updated to v5.14 [#4579]
```
Comment (pllim) [inline CHANGES.rst]: What is this? Seems unrelated.

File: astropy/modeling/fitting.py [hunk 4]
```diff
@@ -30,6 +30,7 @@
 import warnings
 
 from functools import reduce
+from pkg_resources import iter_entry_points
```
Comment (pllim) [inline astropy/modeling/fitting.py]: Is this an *optional* dependency of Astropy or not? If so, we treat it like we do with Scipy.

File: astropy/modeling/fitting.py [hunk 5]
```diff
@@ -1068,3 +1070,35 @@ def _validate_model(model, supported_constraints):
 
     model_copy = model.copy()
     return model_copy
+
+
+def populate_entry_points(entry_points):
```
Comment (pllim) [inline astropy/modeling/fitting.py]: This seems to me that it could be generalized a bit more and moved to `utils` instead?

File: astropy/modeling/tests/test_fitters.py [hunk 6]
```diff
@@ -22,7 +22,12 @@
 from ...utils.data import get_pkg_data_filename
 from ...tests.helper import pytest
 from .utils import ignore_non_integer_warning
+from ...utils.exceptions import AstropyUserWarning
+from pkg_resources import EntryPoint
+from ..fitting import populate_entry_points
 from ...stats import sigma_clip
+import mock
```
Comment (pllim) [inline astropy/modeling/tests/test_fitters.py]: Another *optional* dependency here?

File: astropy/modeling/fitting.py [hunk 7]
```diff
@@ -1068,3 +1070,35 @@ def _validate_model(model, supported_constraints):
...
+    """
+    This injects entery points into the astropy.modeling.fitting namespace
```
Comment (MSeifert04) [inline astropy/modeling/fitting.py]: Small typo here: "entery".

**Comment by pllim** [comment]
I think the user-facing documentation also needs to be updated to explain how this PR enables Sherpa fitters or something like that. Adding examples would be even better, but they should be excluded from `doctest` because of Sherpa.

File: CHANGES.rst (hunk 3 above)
Comment (nocturnalastro) [inline CHANGES.rst]: Not sure what  happened there perhaps it was in an old Changes.rst and wasn't removed when I merged? (there was some conflicts so I did it locally first to fix them)

**Comment by pllim** [comment]
Also, the commits need to get squashed before merge. FYI.

File: CHANGES.rst (hunk 3 above)
Comment (pllim) [inline CHANGES.rst]: Could be. But regardless, this line needs to be removed from your PR.

File: astropy/modeling/fitting.py (hunk 4 above)
Comment (nocturnalastro) [inline astropy/modeling/fitting.py]: I'm pretty sure this comes with setuptools

Comment (astrofrog) [inline astropy/modeling/fitting.py]: Yes, it comes with setuptools. We don't explicitly list setuptools as a required dependency at the moment though. However, I think we could easily make a case that it should be. It would make sense to send an email to astropy-dev about this.

File: astropy/modeling/tests/test_fitters.py (hunk 6 above)
Comment (astrofrog) [inline astropy/modeling/tests/test_fitters.py]: Yes, mock is definitely an optional dependency. Again, we could probably make the case for including it as a test requirement (since it's actually now part of the standard library in recent versions of Python 3). Actually, maybe it makes sense to have a ``try...execpt`` to try importing from the standard library first.

File: astropy/modeling/fitting.py (hunk 5 above)
Comment (taldcroft) [inline astropy/modeling/fitting.py]: @pllim - we had some discussion about whether to turn this entry points functionality into a generalized utility as you suggest.  In the end we decided that since there is currently only one use case that it would be sensible to leave this as a dedicated function for fitting.  Then when/if there are new use cases for this in other sub-packages it will be easier to make decisions about API.

Comment (hamogu) [inline astropy/modeling/fitting.py]: I agree with @taldcroft here.

File: astropy/modeling/fitting.py [hunk 8]
```diff
@@ -1068,3 +1076,51 @@ def _validate_model(model, supported_constraints):
 
     model_copy = model.copy()
     return model_copy
+
+
+def populate_entry_points(entry_points):
+    """
+    This injects entry points into the astropy.modeling.fitting namespace.
```
Comment (taldcroft) [inline astropy/modeling/fitting.py]: `astropy.modeling.fitting` should have reference backticks.

File: .travis.yml [hunk 9]
```diff
@@ -45,7 +45,10 @@ matrix:
         - os: osx
           env: SETUP_CMD='test'
                CONDA_DEPENDENCIES=$CONDA_ALL_DEPENDENCIES
-               PIP_DEPENDENCIES='jplephem'
+               if [[ $PYTHON_VERSION == 2.7 ]]; 
+                 then PIP_DEPENDENCIES='jplephem mock';
```
Comment (bsipocz) [inline .travis.yml]: this looks a very weird formatting, could you the usual one, putting else on the same level of if?

File: .travis.yml [hunk 10]
```diff
@@ -25,7 +25,7 @@ env:
         - NUMPY_VERSION=stable
         - MAIN_CMD='python setup.py'
         - CONDA_DEPENDENCIES='Cython jinja2'
-        - CONDA_ALL_DEPENDENCIES='Cython jinja2 scipy h5py matplotlib pyyaml scikit-image pandas pytz beautifulsoup4 ipython mpmath'
+        - CONDA_ALL_DEPENDENCIES='Cython jinja2 scipy h5py matplotlib pyyaml scikit-image pandas pytz beautifulsoup4 ipython mpmath setuptools'
```
Comment (bsipocz) [inline .travis.yml]: I think you don't need to list setuptools here

File: astropy/modeling/tests/test_fitters.py [hunk 11]
```diff
@@ -24,12 +23,33 @@
 from .utils import ignore_non_integer_warning
 from ...stats import sigma_clip
 
+from ...utils.exceptions import AstropyUserWarning
+from ..fitting import populate_entry_points
+import warnings
+
+from pkg_resources import EntryPoint
```
Comment (cdeil) [inline astropy/modeling/tests/test_fitters.py]: Did you mean to have `from pkg_resources import EntryPoint` here?
You have this line again below in a try ... except block.

File: docs/install.rst [hunk 12]
```diff
@@ -44,6 +44,9 @@ Astropy also depends on other packages for optional features:
 
 - `objgraph <https://mg.pov.lt/objgraph/>`_: Used only in tests to test for reference leaks.
 
+- `setuptools <https://pythonhosted.org/setuptools/>`_: Used for dicovery of entry points which are used to insert fitters into modeling.fitting
```
Comment (cdeil) [inline docs/install.rst]: Typo: `dicovery` -> `discovery`

File: astropy/modeling/tests/test_fitters.py (hunk 11 above)
Comment (nocturnalastro) [inline astropy/modeling/tests/test_fitters.py]: No I'll remove it now.

File: docs/install.rst [hunk 13]
```diff
@@ -44,6 +44,9 @@ Astropy also depends on other packages for optional features:
 
 - `objgraph <https://mg.pov.lt/objgraph/>`_: Used only in tests to test for reference leaks.
 
+- `setuptools <https://pythonhosted.org/setuptools/>`_: Used for discovery of entry points which are used to insert fitters into modeling.fitting
+
+- `mock <https://github.com/testing-cabal/mock>`_ (python <= 3.2) or `unittest.mock <https://docs.python.org/dev/library/unittest.mock.html>`_ (python > 3.3): Used for geting the entry points
```
Comment (taldcroft) [inline docs/install.rst]: Should be "Used for testing the entry point functionality in modeling.fitting".

File: astropy/modeling/fitting.py [hunk 14]
```diff
@@ -1068,3 +1075,50 @@ def _validate_model(model, supported_constraints):
...
+
+    """
+
+    for entry_point in entry_points:
+        name = entry_point.name
+        try:
+            entry_point = entry_point.load()
+        except Exception as e:
+            # This stops the fitting from choking if an entry_point produces an error.
+            warnings.warn(AstropyUserWarning('{type} error occured in entry '
```
Comment (taldcroft) [inline astropy/modeling/fitting.py]: typo: `occurred`

**Comment by taldcroft** [comment]
I think the main thing remaining here from my perspective is a new subsection within http://astropy.readthedocs.io/en/latest/modeling/new.html#defining-new-fitter-classes that might be `Defining a Plugin Fitter`.

This does not need to be very long, but just explain the basic concept and show the minimal additional setup and code required.  This would just come straight from Saba.  It can also contain a link to the Saba RTD as a reference example.

While you are mucking about with narrative docs, it might be worth putting a short section `Plugin Fitters` into http://astropy.readthedocs.io/en/latest/modeling/fitting.html which says roughly the same thing (i.e. "There is a plugin mechanism set up to allow automatic insertion of external Fitter classes into the astropy.fitting namespace").  Then mention that the current existing package which takes advantage of this is Saba which allows fitting using the Sherpa fitting engine.  So really just a few sentences here, but we do want users to be able to discover Saba from astropy.modeling.  But this should not provide any specific examples of usage in order to keep the package distinction clear (for licensing purposes).

**Comment by nocturnalastro** [comment]
@taldcroft  is this ok?

File: docs/modeling/fitting.rst [hunk 15]
```diff
@@ -163,3 +163,31 @@ bounds internally.
...
     >>> fitting.SLSQPLSQFitter.supported_constraints
     ['bounds', 'eqcons', 'ineqcons', 'fixed', 'tied']
+
+Plugin Fitters
+--------------
+
+
+Fitters defined outside of astropy's core can be inserted into the
+`astropy.modeling.fitting` namespace through the use of entry points. 
+Entry points are references to importable objects a tutorial on
```
Comment (taldcroft) [inline docs/modeling/fitting.rst]: `to importable object and a tutorial ..`

File: docs/modeling/fitting.rst [hunk 16]
```diff
@@ -163,3 +163,31 @@ bounds internally.
...
     ['bounds', 'eqcons', 'ineqcons', 'fixed', 'tied']
+
+Plugin Fitters
+--------------
+
+
+Fitters defined outside of astropy's core can be inserted into the
+`astropy.modeling.fitting` namespace through the use of entry points. 
+Entry points are references to importable objects a tutorial on 
+defining entry points can be found in `setuptools' documentation
```
Comment (taldcroft) [inline docs/modeling/fitting.rst]:
```
in the `setuptools documentation <...>`_.  # with a period at the end
```

File: docs/modeling/fitting.rst [hunk 17]
```diff
@@ -163,3 +163,31 @@ bounds internally.
...
+    setup(
+          # ...
+          entry_points = {'astropy.modeling': 'PluginFitterName = fitter_module:PlugFitterClass'}
+    )
+
+This would allow users to import the `PlugFitterName` through `astropy.modeling.fitting` by::
+  
+    from astropy.modeling.fitting import PlugFitterName
+
+One project which uses this functionality is `Saba <https://saba.readthedocs.io/>`_
```
Comment (taldcroft) [inline docs/modeling/fitting.rst]: Put a comma after the `Saba` link, then replace `to insert its` with `which inserts its`.  Then change `which allows` to `class and thus allows`.

**Comment by hamogu** [comment]
@nocturnalastro : While all of @taldcroft 's comments makes sense, I should point out that the mere fact that you get comments on this level means that this PR could be merged right now.

File: docs/modeling/new.rst [hunk 18]
```diff
@@ -327,6 +327,30 @@ necessary::
                                              ineqcons=self.ineqcons)
         return model_copy
 
+Defining a Plugin Fitter
+------------------------
+
+The is now a plugin mechanism which allows fitters defined outside of
```
Comment (taldcroft) [inline docs/modeling/new.rst]: In general one should avoid time references like "now" in docs because in 2 or 10 years that doc line will still probably be there.

So say "Astropy.modeling includes a plugin .."

File: docs/modeling/new.rst [hunk 19]
```diff
@@ -327,6 +327,30 @@ necessary::
...
+ astropy's core can be inserted into the`astropy.modeling.fitting`
```
Comment (taldcroft) [inline docs/modeling/new.rst]: Replace "can be" with "to be".

File: docs/modeling/fitting.rst [hunk 20]
```diff
@@ -163,3 +163,31 @@ bounds internally.
...
+Plugin Fitters
+--------------
+
+
+Fitters defined outside of astropy's core can be inserted into the
+`astropy.modeling.fitting` namespace through the use of entry points. 
+Entry points are references to importable objects a tutorial on 
+defining entry points can be found in `setuptools' documentation 
+<http://setuptools.readthedocs.io/en/latest/setuptools.html#dynamic-discovery-of-services-and-plugins>`_
+Plugin fitters are required to extend from the `~astropy.modeling.fitting.Fitter`
```
Comment (taldcroft) [inline docs/modeling/fitting.rst]: Change "are required" to "must".  The "are required" wording is ambiguous and can mean that one must use plugin fitters in order to extend the fitter class.

File: docs/modeling/new.rst [hunk 21]
```diff
@@ -327,6 +327,30 @@ necessary::
...
 
+Defining a Plugin Fitter
+------------------------
+
+The is now a plugin mechanism which allows fitters defined outside of
+ astropy's core can be inserted into the`astropy.modeling.fitting` 
+namespace through the use of entry points. Entry points are 
+references to importable objects a tutorial on defining entry points 
+can be found in `setuptools' documentation <http://setuptools.readthedocs.io/en/latest/setuptools.html#dynamic-discovery-of-services-and-plugins>`_
+Plugin fitters are required to extend from the `~astropy.modeling.fitting.Fitter`
```
Comment (taldcroft) [inline docs/modeling/new.rst]: Apply the comments from the previous section here.

**Comment by taldcroft** [comment]
@nocturnalastro - the new docs are generally OK but I had a number of wording suggestions.  Also it looks like there is a doc build failure that needs to be looked at.  You might need to +SKIP some lines?

**Comment by taldcroft** [comment]
@nocturnalastro - sorry for the long delay, this got away from me.  Looks good but I'm a little confused with the code block sections and the `+SKIP`.  The functionality that I know about is using `# doctest: +SKIP` for each code line to skip.  Is `+SKIP` actually working if you build the docs locally?  The fact that it is passing CI is encouraging, but I just wasn't aware of skipping a whole code block with this syntax.

**Comment by nocturnalastro** [comment]
It isn't stripped either way. 
I'd have to add `>>>` at the beginning  but that is different to the style of all the other cells.

**Comment by mirca** [comment]
@nocturnalastro I'm pretty sure you could use ``.. doctest-skip::`` to skip a block code. Looking forward to see this in, and btw, @nocturnalastro congratulations on your awesome GSoC project!

**Comment by nocturnalastro** [comment]
One of the errors is coming from a `.. doctest-skip::` block.

File: CHANGES.rst [hunk 22]
```diff
@@ -59,6 +59,8 @@ New Features
 
 - ``astropy.wcs``
 
+- Added the injection of EntryPoints into astropy.modeling.fitting if they inherit from Fitters class. [#5241]
```
Comment (taldcroft) [inline CHANGES.rst]: This needs to go into the ``modeling`` section above, and add a line break (probably after ``if`` or ``they``) to match the existing formatting.

**Comment by taldcroft** [comment]
@nocturnalastro - I changed the milestone to 1.3.0.  I originally put it at 1.2.2 but I honestly don't recall why since this is an API enhancement and not a bug fix.

Otherwise just the one comment on the CHANGES.rst entry.  You can `[skip ci]` that and I will check for correctness and then merge.

**Comment by taldcroft** [comment]
@nocturnalastro - sorry, you have a merge conflict now in CHANGES.rst and this needs to be rebased.  The good news is that the rebase should be easy (I just tried it).

**Comment by nocturnalastro** [comment]
No worries, should I squash the commits while I'm at it?

**Comment by nocturnalastro** [comment]
:confused: I did a rebase and it's still conflicting.
@taldcroft  any clue?

**Comment by taldcroft** [comment]
Did you rebase on the very latest master?  Squashing commits might be a good thing in this case.

**Comment by nocturnalastro** [comment]
It had just pulled down the master in my rep.

**Comment by taldcroft** [comment]
Congratulations @nocturnalastro, merged!  Thanks again for a great GSoC project.
